### ISSUE TITLE
max.poll.interval.ms is not respected anymore

### ISSUE THREAD

**Issue by redbaron** [fede1024/rust-rdkafka#264]
I think that https://github.com/fede1024/rust-rdkafka/pull/252/files  broke `max.poll.interval.ms` handling.

Purpose of that parameter is to leave group if processing takes too long. Processing time is defined as interval between calls to `client.poll`. 

If we look at official Java client, there is a [background thread](https://github.com/apache/kafka/blob/2.5.0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L1262) sending heartbeats, these should be sent at least every `session.timeout.ms` ms. Then there is consumer group coordinator and fetch client, these have their own respective `poll` methods,  **both methods are called synchronously** when application calls `client.poll()`.

`ConsumerCoordinator.poll` updates last poll time, which prevents heartbeat thread from initiating ejection from a group.

In other words **app code needs to call poll often enough to stay in the group**.

In librdkafka, from what I can see, there is a similar setup. There is a periodic heartbeat running from background thread. It also uses last poll time to decide whether to leave group. In librdkafka last poll time **is set from rd_kafka_poll** call which in current rust-rdkafka master is not tied to application level polls, thus preventing client from ever leaving the group because of processing taking to long.

**Comment by benesch** [comment]
The `max.poll.interval.ms` parameter has never applied quite right to the `StreamConsumer`. The point of the `StreamConsumer` that it launches a thread to manage polling on your behalf. It's true that the old code would stop calling poll when the channel buffer filled up, but it would also continually call poll (preventing `max.poll.interval.ms` from ever getting exceeding) if the consumer was not producing any data. This could legitimately hide moments where the processing of the data has gotten stuck, e.g., imagine if you wedged after pulling the last few records out of a topic—it would look like you were still calling poll, even though you'd wedged. 

Ugh. I'm not sure whether it is better to preserve the spirit of the existing `StreamConsumer` (i.e., never exceeding `max.poll.interval.ms`), or to change the behavior to better match the rest of the Kafka ecosystem.

**Comment by redbaron** [comment]
>  The point of the StreamConsumer that it launches a thread to manage polling on your behalf.

I see point of StreamConsumer is to get futures::Stream , nothing else. How it is achieved is implementation detail.

>  This could legitimately hide moments where the processing of the data has gotten stuck,

For this to be left undetected it must be a rare coincidence that app thread wedges AND there are no more messages in the topic. Although not ideal, it is still better than current master, where there is no chance of detecting it ever.

> Ugh. I'm not sure whether it is better to preserve the spirit of the existing StreamConsumer (i.e., never exceeding max.poll.interval.ms), or to change the behavior to better match the rest of the Kafka ecosystem.

Whole rust-rdkafka is "application" side of kafka client. If there was no background thread or tasks and we rely only on rust app code to come back and fetch message in a timely fashion, then we wont break any contracts and stay consistent with rest of Kafka.

[KIP-62](https://cwiki.apache.org/confluence/display/KAFKA/KIP-62%3A+Allow+consumer+to+send+heartbeats+from+a+background+thread) specifies motivation behind `max.poll.interval.ms` as a mechanism to detect wedging without sacrificing time to detect hard client failures. To achieve this goal polls **must** be synchronous with message processing on app level, this matches nicely Rust Future model, so why deviate from it?

**Comment by benesch** [comment]
> I see point of StreamConsumer is to get futures::Stream , nothing else. How it is achieved is implementation detail.

The docs for the `StreamConsumer` in the current edition say this:

> This consumer doesn't need to be polled since it has a separate polling thread.

So there is a bit of commitment to a particular implementation.

> so why deviate from it?

Because perhaps there are existing users of `StreamConsumer` who prefer the current behavior. I only have two data points: that both behaviors are fine for my use case (we just disable the max poll timeout), and that you want the `max.poll.interval.ms` behavior. That's not a lot of data.

If @fede1024 is around, perhaps he can clarify what the intended behavior of the `StreamConsumer` was, and whether he thinks moving to an implementation that allows `max.poll.interval.ms` to apply makes sense.

**Comment by redbaron** [comment]
>> This consumer doesn't need to be polled since it has a separate polling thread.

It is "[Unlike base consumer,] this consumer...". It communicates that there is no need for busy looping in app code, opposite to what BaseConsumer wants you to do. Mention of the thread just clarifies how "no need to be polled" is achieved, but is not main point of this message. "polled" here is `BaseConsumer::poll`, not "Future::poll` , because obviously it still need to be future-polled.

> Because perhaps there are existing users of StreamConsumer who prefer the current behavior. 
> That's not a lot of data.

I'd like to think existing users prefer to have compliant Kafka client, threads or not.
Latest released version behaviour is more or less in line with official Java client. Current master behaviour make `max.poll.interval.ms` parameter moot.

IMHO preserving behaviour of `max.poll.interval.ms` is must,  it remains important part of Kafka dynamic consumer  group membership management, so I suggest either move back to background thread or find another way to respect `max.poll.interval.ms`
